### DISCUSSION TITLE
Welcome to nextjs-boilerplate Discussions!

### DISCUSSION THREAD

**Discussion by NoahDuongMaster** [NoahDuongMaster/vibe-code-stack-for-ceos#4] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by vienlamthe** [comment]
Em chào anh, em đang là sinh viên IT, em đang tìm hiểu về NextJS và biết đến repo của anh. Em có 1 câu hỏi liên quan đến cách triển khai feat đa ngôn ngữ ạ. 

Em có thử triển khai feat theo đa ngôn ngữ theo docs thì bình thường nhưng em muốn phải đồng bộ lên URL. Em xem qua demo repo của anh thì thấy anh không đồng bộ lên url mà thay vào đó anh lưu key lang vào cookies và truy xuất data ngôn ngữ ở hook. Em cũng thử nhưng bị 1 lỗi là Error hydrate thì em hiểu là việc render giữa server và client đang khác nhau. Em lại mò mẫm repo anh thì anh có xử lý phần hydrate ở _app.tsx . Em hiểu là anh muốn render phần locales sau lần render đầu tiên. Nhưng em lại thấy anh check đoạn code này:

```
{hydated &&
                getLayout(
                  <AppAntd notification={{ placement: 'topRight' }}>
                    <Component {...pageProps} />
                  </AppAntd>,
                )}
```

Em nghĩ chỗ xử lý này đã biến các component này mất đi tính chất server-rendering của Nextjs và performance nó không tốt. Em giải thích qua chữ thì có lẽ sẽ khó hiểu và không chính xác nên mong anh thông cảm ạ. Cảm ơn anh đã xem và chia sẻ repo này đến với cộng đồng. Respect!

**Reply by NoahDuongMaster** [reply]
@vienlamthe Hi em, 
Trước khi trả lời câu hỏi của em, anh muốn cảm ơn em vì đã để lại comment góp ý này nhé.

Việc làm đa ngôn ngữ mà tạo thành nhiều prefix trên url thì quá đơn giản rồi, như em có nói là vào xem docs là làm được, còn với Nextjs thì hiện tại anh chưa thấy có package nào làm được nên anh phải tự viết function cho riêng mình.

Nhưng do anh không muốn tạo thành prefix url, anh chỉ muốn nó đồng nhất 1 url thôi, khi mà người dùng đổi ngôn ngữ thì sẽ lưu lại vào cookie => Lần sau khi họ vào lại website thì sẽ hiển thị lại ngồi ngữ lần trước họ đã chọn.

**Đây tuỳ chọn dựa trên nhu cầu của mỗi người, em có thể sửa thành như docs nếu em thích nhé.**
> Em xem qua demo repo của anh thì thấy anh không đồng bộ lên url mà thay vào đó anh lưu key lang vào cookies và truy xuất data ngôn ngữ ở hook.

Nó không hẳn là mất tính chất của SSR em. SRR vẫn hoạt động bình thường. Lý do chỉ là do mình đang sử dụng thư viện UI là Antd Design mà hiện tại thư viện này chưa support hoàn toàn SSR như MUI nên a làm như vậy với ý đồ đợi cho CSS load hết xong mới hiện ra UI tương tự như loading đó. Không có vấn đề gì nghiêm trọng hay mất đi tính SSR đâu em.

Nếu em không làm như anh thì khi mà dự án to lên, nhiều CSS nặng dần lên thì e sẽ thấy hiện tượng trang web bị nháy nháy, giật giật. Anh đã làm nhiều dự án thực tế rồi và bị khách hàng họ bắt bug nên anh cũng sửa luôn vào đây.
> Em lại mò mẫm repo anh thì anh có xử lý phần hydrate ở _app.tsx . Em hiểu là anh muốn render phần locales sau lần render đầu tiên. Nhưng em lại thấy anh check đoạn code này:

**Reply by NoahDuongMaster** [reply]
@vienlamthe Em có thể cùng anh đóng góp cho dự án này em nhé. Nếu em phát hiện thấy bất cứ bug hay issue gì thì e có thể vào phần issues rồi tạo issue. Sau đó mình sẽ cùng nhau trao đổi.

Em có thể tạo PR đẩy lên cho anh merge nhé.

Thank em.

**Reply by vienlamthe** [reply]
Em cảm ơn anh đã trả lời thắc mắc của em. Em rất muốn đóng góp vào project này nhưng không biết bên anh có plan sắp tới chưa ạ? Nếu có thì em xin 1 slot với tinh thần học hỏi cũng như có cơ hội được va chạm bugs :)) Tech stack của em: Vuejs, Reactjs, Nodejs, Nestjs ạ

**Reply by NoahDuongMaster** [reply]
@vienlamthe Hi em,

Anh có 2 cái issues này nè. Em có thể làm rồi tạo PR cho anh.

<img width="1224" alt="image" src="https://github.com/truongdn-it/nextjs-boilerplate/assets/41463958/428c6d08-6b63-467f-8c37-0ada8e324cba">

**Reply by vienlamthe** [reply]
Em chào anh, em thấy có 1 vấn đề nữa ở chỗ feat **Đa ngôn ngữ**. Khi reload page thì các text được translate sẽ bị nháy nháy. Kiểu như nó render 'key' file common.json trước rồi mới render ra text hoàn chỉnh. Em tham khảo một vài trang khác thì không thấy tình trạng nháy như thế. Issue này anh có solution thế nào ạ? Thank anh.

**Reply by NoahDuongMaster** [reply]
@vienlamthe Hi em,

Issue này không cần giải quyết em nhé. Vì đơn giản nó là client-side em. Nên nó sẽ có hiện tượng như vậy. Đó là do mình lựa chọn là sử dụng đa ngôn ngữ ở phía CSR. Cái này nhiều trang cũng bị như mình mà.

Nếu sửa thì sẽ làm theo cách sau:
- Dưới SSR get trước cookie xem ngôn ngữ đang là gì, rồi trả về text đa ngôn ngữ trước luôn là được.
- Trên CSR thì truyền cục text đa ngôn ngữ đó vào làm giá trị khởi tạo.

Nói chung là fix được nhưng không ưu tiên em.
